http/src/main/java/io/micronaut/http/filter/FilterRunner.java

```diff
@@ -420,26 +161,72 @@ public final FilterRunner reactorContext(Context reactorContext) {
      * response
      */
     public final ExecutionFlow<MutableHttpResponse<?>> run(HttpRequest<?> request) {
-        if (this.request != null) {
-            throw new IllegalStateException("Can only process one request");
-        }
-        this.request = request;
+        return (ExecutionFlow) filterRequest(new FilterContext(request, initialReactorContext), filters.listIterator(), new HashMap<>());
+    }
 
-        ExecutionFlow<HttpResponse<?>> resultFlow = CompletableFutureExecutionFlow.just(responseSuspensionPoint);
-        workRequest();
-        //noinspection unchecked,rawtypes
-        return (ExecutionFlow) resultFlow;
+    private ExecutionFlow<HttpResponse<?>> filterRequest(FilterContext context,
+                                                         ListIterator<GenericHttpFilter> iterator,
+                                                         Map<GenericHttpFilter, Map.Entry<ExecutionFlow<FilterContext>, FilterContinuationImpl<?>>> suspended) {
+        return filterRequest0(context, iterator, suspended)
+                .flatMap(newContext -> {
+                    if (newContext.response != null) {
+                        return filterResponse(newContext, iterator, null, suspended);
+                    }
+                    return ExecutionFlow.error(new IllegalStateException("Request filters didn't produce any response!"));
+                });
     }
 
-    private void workRequest() {
-        while (true) {
-            if (!workRequestFilter(filters.get(index++))) {
-                return;
-            }
+    private ExecutionFlow<FilterContext> filterRequest0(FilterContext context,
+                                                        ListIterator<GenericHttpFilter> iterator,
+                                                        Map<GenericHttpFilter, Map.Entry<ExecutionFlow<FilterContext>, FilterContinuationImpl<?>>> suspended) {
+        if (context.response != null) {
+            return ExecutionFlow.just(context);
+        }
+        if (iterator.hasNext()) {
+            GenericHttpFilter filter = iterator.next();
+            return processRequestFilter(filter, context, suspended)
+                    .onErrorResume(throwable -> {
+                        // Un-suspend possibly awaiting filter and exception filtering scenario of the http client
+                        return filterResponse(context, iterator, throwable, suspended).map(context::withResponse);
+                    })
+                    .flatMap(newContext -> filterRequest0(newContext, iterator, suspended));
```

Comment:
@dstepanov did you address this feedback?

Reply:
Moved the method before the error handle to skip ot

Reply:
i think that is still wrong, since now onErrorResume can be called after all the other response filters run, if there is another response filter that fails

Reply:
If all the response filters are run, nothing will be run. It will continue the response flow on error, but now it will only trigger error filters.

Reply:
nothing will be run because the iterator will be back at the start in onErrorResume, correct? it's a bit confusing but i think i understand